src/tokens/tokens.service.ts

```diff
@@ -0,0 +1,398 @@
+    const response = await lastValueFrom(
+      this.http.post<EthConnectAsyncResponse>(
+        `${this.contractInstanceUrl}/${dto.poolId}/burnWithData`,
+        {
+          data: encodeHex(dto.data ?? ''),
+          from: dto.from,
+          amount: dto.amount,
+        },
+        this.postOptions(dto.operator, dto.requestId),
+      ),
+    );
+    return { id: response.data.id };
+  }
+
+  async balance(dto: TokenBalanceQuery): Promise<TokenBalance> {
+    const response = await lastValueFrom(
+      this.http.get<EthConnectReturn>(`${this.contractInstanceUrl}/${dto.poolId}/balanceOf`, {
+        params: {
+          account: dto.account,
+        },
+        ...basicAuth(this.username, this.password),
+      }),
+    );
+
+    return { balance: response.data.output };
+  }
+
+  async getReceipt(id: string): Promise<EventStreamReply> {
+    const response = await lastValueFrom(
+      this.http.get<EventStreamReply>(`${this.baseUrl}/reply/${id}`, {
+        validateStatus: status => status < 300 || status === 404,
+        ...basicAuth(this.username, this.password),
+      }),
+    );
+    if (response.status === 404) {
+      throw new NotFoundException();
+    }
+    return response.data;
+  }
+}
+
+class TokenListener implements EventListener {
+  private readonly logger = new Logger(TokenListener.name);
+
+  private uriPattern: string | undefined;
+
+  constructor(
+    private http: HttpService,
+    private instanceUrl: string,
+    private topic: string,
+    private username: string,
+    private password: string,
+  ) {}
+
+  async onEvent(subName: string, event: Event, process: EventProcessor) {
+    switch (event.signature) {
+      case tokenCreateEventSignature:
+        process(this.transformTokenCreateEvent(subName, event));
+        break;
+      case transferEventSignature:
+        process(await this.transformTransferEvent(subName, event));
+        break;
+      default:
+        this.logger.error(`Unknown event signature: ${event.signature}`);
+        return undefined;
+    }
+  }
+
+  private transformTokenCreateEvent(
+    subName: string,
+    event: TokenCreateEvent,
+  ): WebSocketMessage | undefined {
+    const { data } = event;
+    const decodedData = decodeHex(data.data ?? '');
+
+    return {
+      event: 'token-pool',
+      data: <TokenPoolEvent>{
+        standard: TOKEN_STANDARD,
+        poolId: event.data.contract_address,
+        type: TokenType.FUNGIBLE,
+        operator: data.operator,
+        data: decodedData,
+        timestamp: event.timestamp,
+        rawOutput: data,
+        transaction: {
+          blockNumber: event.blockNumber,
+          transactionIndex: event.transactionIndex,
+          transactionHash: event.transactionHash,
+          logIndex: event.logIndex,
+          signature: event.signature,
+        },
+      },
+    };
+  }
+
+  private transformTransferEvent(
+    subName: string,
+    event: TransferEvent,
+    eventIndex?: number,
```

Comment:
I don't think you ever pass `eventIndex` do you? It was in ERC1155 for handling some things around batched transfers. If it's unused, just remove it and any processing of it.